engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/NadelServiceTypeFilterTransform.kt

```diff
@@ -86,17 +86,18 @@ class NadelServiceTypeFilterTransform : NadelTransform<State> {
             service.name == IntrospectionService.name -> return null
         }
 
-        val typeNamesOwnedByService = getTypeNamesServiceOwns(executionBlueprint, service)
+        val typeNamesOwnedByService = executionBlueprint.getOverAllTypeNamesForService(service)
         // Add underlying type names as well to handle combination of hydration and renames.
         // Transforms are applied to hydration fields as well, and those fields always reference
         // elements from the underlying schema
-        val underlyingTypeNamesOwnedByService =
-            getUnderlyingTypeNamesServiceOwns(executionBlueprint, service, typeNamesOwnedByService)
-
-        val allTypeNamesOwnedByService = typeNamesOwnedByService + underlyingTypeNamesOwnedByService
+        val underlyingTypeNamesOwnedByService = executionBlueprint.getUnderlyingTypeNamesForService(service)
 
         val fieldObjectTypeNamesOwnedByService = overallField.objectTypeNames
-            .filter { it in allTypeNamesOwnedByService }
+            .filter {
+                // it is MUCH quicker to compare membership in 2 sets rather than
+                // concat 1 giant set and then check
+                it in typeNamesOwnedByService || it in underlyingTypeNamesOwnedByService
+            }
```

Comment:
2 `in` calls are significantly cheaper than ` allTypeNamesOwnedByService = typeNamesOwnedByService + underlyingTypeNamesOwnedByService`